src/compile/data/assemble.ts

```diff
@@ -0,0 +1,242 @@
+import {every, vals} from '../../util';
+import {VgData} from '../../vega.schema';
+import {AggregateNode} from './aggregate';
+import {BinNode} from './bin';
+import {DataFlowNode, OutputNode} from './dataflow';
+import {FacetNode} from './facet';
+import {ParseNode} from './formatparse';
+import {NonPositiveFilterNode} from './nonpositivefilter';
+import {NullFilterNode} from './nullfilter';
+import * as optimizers from './optimizers';
+import {optimizeFromLeaves} from './optimizers';
+import {OrderNode} from './pathorder';
+import {SourceNode} from './source';
+import {StackNode} from './stack';
+import {TimeUnitNode} from './timeunit';
+import {CalculateNode, FilterNode} from './transforms';
+
+/**
+ * Start optimization path from the root. Useful for removing nodes.
+ */
+function optimizeFromRoots(node: DataFlowNode) {
+  // remove empty non positive filter
+  if (node instanceof NonPositiveFilterNode && every(vals(node.filter), b => b === false)) {
+    node.remove();
+  }
+
+  // remove empty null filter nodes
+  if (node instanceof NullFilterNode && every(vals(node.filteredFields), f => f === null)) {
+    node.remove();
+  }
+
+  // remove output nodes that are not required
+  if (node instanceof OutputNode && !node.required) {
+    node.remove();
+  }
+
+  node.children.forEach(optimizeFromRoots);
+}
+
+
+/**
+ * Return all leaf nodes.
+ */
+function getLeaves(roots: DataFlowNode[]) {
+  const leaves: DataFlowNode[] = [];
+  function append(node: DataFlowNode) {
+    if (node.numChildren() === 0) {
+      leaves.push(node);
+    } else {
+      node.children.forEach(append);
+    }
+  }
+
+  roots.forEach(append);
+  return leaves;
+}
+
+/**
+ * Print debug information for dataflow tree.
+ */
+function debug(node: DataFlowNode) {
+  console.log(`${(node.constructor as any).name}${node.debugName ? ` (${node.debugName})` : ''} -> ${
+    (node.children.map(c => {
+      return `${(c.constructor as any).name}${c.debugName ? ` (${c.debugName})` : ''}`;
+    }))
+  }`);
+  console.log(node);
+  node.children.forEach(debug);
+}
+
+function makeWalkTree(data: VgData[]) {
+  // to name datasources
+  let datasetIndex = 0;
+
+  /**
+   * Recursively walk down the tree.
+   */
+  function walkTree(node: DataFlowNode, dataSource: VgData) {
+    if (node instanceof ParseNode) {
+      if (node.parent instanceof SourceNode && dataSource.format) {
+        dataSource.format.parse = node.assemble();
+      } else {
+        throw new Error('Can only instantiate parse next to source.');
+      }
+    }
+
+    if (node instanceof FacetNode) {
+      if (!dataSource.name) {
+        dataSource.name = `data_${datasetIndex++}`;
```

Comment:
I assume that this is a faceted data source that faceted marks will refer to?  If so, name it `faceted_`?